connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java

```diff
@@ -238,6 +261,11 @@ private static Object castValueToType(Schema schema, Object value, Schema.Type t
             // Ensure the type we are trying to cast from is supported
             validCastType(inferredType, FieldType.INPUT);
 
+            // Perform logical type casts to their internal representation.
+            if (schema != null && targetType != Type.STRING && LOGICAL_TYPES.contains(schema.name())) {
+                value = castLogicalType(schema, value);
```

Comment:
Rather than check whether the `LOGICAL_TYPES.contains(schema.name())`, would it be sufficient to simply do the following:
```
if (schema != null && targetType != Type.STRING) {
    value = castToPrimitive(schema, value);
}
```
and then change the `castLogicalType(...)` to:
```
    protected static Object castToPrimitive(Schema schema, Object value) {
        switch (schema.name()) {
            case Date.LOGICAL_NAME:
                return Date.fromLogical(schema, (java.util.Date) value);
            case Time.LOGICAL_NAME:
                return Time.fromLogical(schema, (java.util.Date) value);
            case Timestamp.LOGICAL_NAME:
                return Timestamp.fromLogical(schema, (java.util.Date) value);
            case Decimal.LOGICAL_NAME:
                return Decimal.fromLogical(schema, (java.math.BigDecimal) value);
        }
        throw new DataException("Casting from " + schema.name() + " to a primitive is not supported");
    }
```
Also, if these methods were not static, then somebody could extend this class with their own customizations.

WDYT?

Reply:
I had gone pretty far down that path of introducing a LogicalType interface with `fromLogical` and `toLogical` methods on them, having these classes implement the non-static methods, what not. Then realized that these are part of the API and used in places like KSQL and schema registry. Also, there could be external clients depending on these static methods that might break if we changed it.

That said, we could do something like have separate instance methods with existing static methods forwarding to instance methods on a singleton.